File: output_log/claude-opus/sample_2/c/qwen3/c/Struct/deposit.c
#include "stdlib.h"

/*@ predicate account(struct account *p; int balance) =
    p != 0 &*&
    malloc_block_account(p, sizeof(struct account)) &*&
    p->balance |-> balance;
@*/

struct account
{
    int balance;
};

struct account *create_account()
//@ requires true;
//@ ensures account(result, 0);
{
    struct account *myAccount = malloc(sizeof(struct account));
    if (myAccount == 0)
    {
        abort();
    }
    //@ assume myAccount != 0;
    //@ leak malloc_block_account(myAccount, sizeof(struct account));
    myAccount->balance = 0;
    //@ close account(myAccount, 0);
    return myAccount;
}

void account_deposit(struct account *myAccount, int amount)
//@ requires account(myAccount, ?b);
//@ ensures account(myAccount, b + amount);
{
    //@ open account(myAccount, ?b);
    myAccount->balance += amount;
    //@ close account(myAccount, b + amount);
}

int account_get_balance(struct account *myAccount)
//@ requires account(myAccount, ?b);
//@ ensures account(myAccount, b) &*& result == b;
{
    //@ open account(myAccount, ?b);
    int result = myAccount->balance;
    //@ close account(myAccount, b);
    return result;
}

void account_set_balance(struct account *myAccount, int newBalance)
//@ requires account(myAccount, ?_);
//@ ensures account(myAccount, newBalance);
{
    //@ open account(myAccount, ?_);
    myAccount->balance = newBalance;
    //@ close account(myAccount, newBalance);
}

void account_dispose(struct account *myAccount)
//@ requires account(myAccount, ?_);
//@ ensures true;
{
    //@ open account(myAccount, ?_);
    free(myAccount);
}

int main()
//@ requires true;
//@ ensures true;
{
    struct account *myAccount = create_account();
    account_set_balance(myAccount, 5);
    account_deposit(myAccount, 10);
    int b = account_get_balance(myAccount);
    assert(b == 15);
    account_dispose(myAccount);
    return 0;
}

File: output/qwen3/rust/purely_unsafe/solutions/deposit.rs
use std::alloc::{alloc, dealloc, handle_alloc_error, Layout};

#[pred account_pred(p: *mut Account) = 
    p != 0 &*& 
    malloc_block_Account(p, 1) &*& 
    *p |-> Account { balance: ?b } &*& 
    acc(_balance_field(p), b)
]

#[pred _balance_field(p: *mut Account) = 
    p != 0 &*& 
    *p |-> Account { balance: ?b } &*& 
    acc(_balance_field(p), b)
]

#[lemma]
fn account_balance_access(p: *mut Account)
    requires account_pred(p)
    ensures account_pred(p) &*& exists(?b) &*& *p |-> Account { balance: b } &*& acc(_balance_field(p), b)
{
    // This lemma is trivial in VeriFast due to the structure of the predicate.
}

impl Account {
    unsafe fn create() -> *mut Account {
        #[requires true]
        #[ensures account_pred(result)]
        let my_account = alloc(Layout::new::<Account>()) as *mut Account;
        if my_account.is_null() {
            handle_alloc_error(Layout::new::<Account>());
        }
        #[requires my_account != 0 &*& malloc_block_Account(my_account, 1)]
        #[ensures account_pred(my_account) &*& *my_account |-> Account { balance: 0 }]
        (*my_account).balance = 0;
        return my_account;
    }

    unsafe fn get_balance(my_account: *mut Account) -> i32 {
        #[requires account_pred(my_account)]
        #[ensures account_pred(my_account) &*& result == (*my_account).balance]
        return (*my_account).balance;
    }

    unsafe fn set_balance(my_account: *mut Account, new_balance: i32) {
        #[requires account_pred(my_account)]
        #[ensures account_pred(my_account) &*& (*my_account).balance == new_balance]
        (*my_account).balance = new_balance;
    }

    unsafe fn deposit(my_account: *mut Account, amount: i32) {
        #[requires account_pred(my_account)]
        #[ensures account_pred(my_account) &*& (*my_account).balance == (*my_account).balance + amount]
        (*my_account).balance += amount;
    }

    unsafe fn dispose(my_account: *mut Account) {
        #[requires account_pred(my_account)]
        #[ensures true]
        dealloc(my_account as *mut u8, Layout::new::<Account>());
    }
}

fn main() {
    unsafe {
        let my_account = Account::create();
        Account::set_balance(my_account, 5);
        Account::deposit(my_account, 10);
        let b = Account::get_balance(my_account);
        std::hint::assert_unchecked(b == 15);
        Account::dispose(my_account);
    }
}